engine/access/state_stream/backend/event_provider.go

```diff
@@ -0,0 +1,175 @@
+package backend
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/model/access"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module/executiondatasync/optimistic_sync"
+	"github.com/onflow/flow-go/storage"
+	"github.com/onflow/flow-go/utils/logging"
+)
+
+// EventsResponse represents the response containing events for a specific block.
+type EventsResponse struct {
+	BlockID          flow.Identifier
+	Height           uint64
+	Events           flow.EventsList
+	BlockTimestamp   time.Time
+	ExecutorMetadata access.ExecutorMetadata
+}
+
+// EventsProvider retrieves events by block height. It can be configured to retrieve events from
+// the events indexer(if available) or using a dedicated callback to query it from other sources.
+type EventsProvider struct {
+	log              zerolog.Logger
+	headers          storage.Headers
+	execDataProvider ExecutionDataProvider
+
+	fetchFromLocalCache bool
+	execResultProvider  optimistic_sync.ExecutionResultInfoProvider
+	execStateCache      optimistic_sync.ExecutionStateCache
+}
+
+func NewEventsProvider(
+	log zerolog.Logger,
+	headers storage.Headers,
+	execDataProvider ExecutionDataProvider,
+	fetchFromLocalCache bool,
+	execResultProvider optimistic_sync.ExecutionResultInfoProvider,
+	execStateCache optimistic_sync.ExecutionStateCache,
+) *EventsProvider {
+	return &EventsProvider{
+		log:                 log,
+		headers:             headers,
+		execDataProvider:    execDataProvider,
+		fetchFromLocalCache: fetchFromLocalCache,
+		execResultProvider:  execResultProvider,
+		execStateCache:      execStateCache,
+	}
+}
+
+// Events return events for a given block height.
+//
+// Expected error returns during normal operations:
+//   - [codes.NotFound]: If a block header for the specified block height is not found.
```

Comment:
this isn't accurate anymore. please update to list the sentinels returned. these should be either sentinels returned directly by this function, or [documented] sentinels returned from functions called within the function